### ISSUE TITLE
Not all domain hosts support wildcard subdomains

### ISSUE THREAD

**Issue by mathiasm74** [githubsaturn/captainduckduck#203]
Apparently not all domain hosts allow use of wildcard subdomain (see https://www.dynadot.com/community/forums/f7-domain-name-help/wildcard-subdomains-9923.html)

Does this mean I have to change domain host, or is there a way around this? 
Can CDD be improved to also work without wildcard subdomains, or is this a non-solvable issue?

**Comment by mathiasm74** [comment]
How common might issue be?

**Comment by githubsaturn** [comment]
https://serverfault.com/questions/236705/domain-registrars-supporting-wildcard-cname-records

If they don't support this basic feature, they are not a decent DNS host.

Having said that, I'll be allowing the advanced users to override some checks in the future, it will be done as part of this issue: https://github.com/githubsaturn/captainduckduck/issues/181

Once that's implemented, you can technically bypass the wildcard check and manually assign some fixed domain records to your DNS entries.
captain.something.com -> 123.123.123.123
registry.something.com -> 123.123.123.123
myapp.something.com -> 123.123.123.123
Of course, this will be tedious, but that's what you have to do if you want to go with second-tier providers. Basic providers such as GoDaddy, NameCheap, and Google provide this functionality.

Even dynadot.com that you shared, somewhat supports wildcard, but up to one level deep. You cannot use a two level deep wildcard such as `*.subdomain.domain.com` using dynadot.

**Comment by drmrbrewer** [comment]
@mathiasm74 if you don't want to change host, something else that you might want to consider is using something like [cloudflare](https://www.cloudflare.com/) to sit between the client and your DNS-deficient host, so that you do your DNS management in cloudflare rather than at your host... and cloudflare definitely *does* support DNS wildcards.  The host still remains as your host, but cloudflare sits in the middle, adding a layer of caching (if you want) to take load off your host and also to improve response times to your client.  But, you can just use it for its DNS features if you like.  It's really easy to set up (and free for personal use)... it was a while ago that I did it, but I recall that it was just a case of specifying your host `mydomain.com` and cloudflare does the rest for you... and you can customise from there.  There are probably services other than cloudflare that do something similar.

**Comment by dkimot** [comment]
Cloudflare does not support wildcard domains below the business level ($200/month). You can still use Cloudflare as a DNS provider, however caching would not be available below the business plan.

**Comment by drmrbrewer** [comment]
@dkimot I think this should be clarified... cloudflare definitely *does* support wildcard domains on the free plan (I am doing exactly this), but not *caching* of wildcard domains.  But if you're not really interested in cloudflare for caching, and just want access to wildcard domains without moving host, it might be something to consider.  For me, traffic to `mydomain.com` is cached by cloudflare (which in itself is a big benefit), but not traffic to `*.something.mydomain.com`... but I don't want it to be anyway.

**Comment by dkimot** [comment]
@drmrbrewer That is what I was trying to get across, I definitely should have clarified the first sentence. Thanks

**Comment by githubsaturn** [comment]
Allowing wildcard to be skipped: https://github.com/githubsaturn/captainduckduck/commit/049a6f0bf74f171ddcef32c01682554e827286c1

Will be released in v0.7.0

**Comment by mxdpeep** [comment]
> Cloudflare does not support wildcard domains below the business level ($200/month). You can still use Cloudflare as a DNS provider, however, caching would not be available below the business plan.

cloudflare supports * for any domains, just not through the proxy service

**Comment by douglasjunior** [comment]
> If they don't support this basic feature, they are not a decent DNS host.

This may be true, but there are still a lot of providers that do not support wildcard. As for example registro.br, which is the official provider of Brazil.

Therefore, we were unable to use a domain in CapRover and therefore HTTPS.

Is there any way to set up a domain to access CapRover without a wildcard and that allows us to use HTTPS?

**Comment by githubsaturn** [comment]
> This may be true, but there are still a lot of providers that do not support wildcard. For example https://registro.br, which is the official provider of ".com.br" domains in Brazil.

Registrar is different from DNS host. You can register your domain with "registro.br" and you can have your DNS hosted by cloudflare. I think [this video](https://www.youtube.com/watch?v=qwyTr0c3l-E) shows how to do it.

Regardless, if you don't want to do this. You can still use CapRover with non-wildcard DNS. To do this, run:
```
echo  "{\"skipVerifyingDomains\":\"true\"}" >  /captain/data/config-override.json
docker service update captain-captain --force
```

Now you just need to map your individual domains to CapRover. i.e,:
```
captain.sub.domain.com.br
myapp.sub.domain.com.br
```

**Comment by douglasjunior** [comment]
Thanks again @githubsaturn !

Unfortunately, we cannot always move the domain to another provider, often the domain is already used for other things and we end up getting stuck in bad providers.

I will test your tips and return with the results. (cc @edvaldoszy)

One trick we did was to create a new empty app on CapRover, and then create a reverse proxy for http://captain-captain:3000.
So we were able to add a domain and activate HTTPS, and then we can close port 3000.
